src/job.ts

```diff
@@ -640,87 +669,118 @@ export class Job {
         return producerReportsEnvs;
     }
 
-    private async copyArtifactsIn(writeStreams: WriteStreams) {
-        if (!this.imageName && !this.shellIsolation) {
-            return;
-        }
+    private async copyCacheIn(writeStreams: WriteStreams) {
+        if (this.mountCache && this.imageName) return;
+        if ((!this.imageName && !this.shellIsolation) || this.cache.length === 0) return;
 
-        if (!this.producers || this.producers.length === 0) {
-            return;
-        }
+        for (const c of this.cache) {
+            if (!["pull", "pull-push"].includes(c.policy)) return;
 
-        const safeJobName = this.safeJobName;
-
-        const cpFunc = async (folder: string) => {
-            if (!this.imageName && this.shellIsolation) {
-                return Utils.spawn(`rsync -a ${folder}/. ${this.cwd}/.gitlab-ci-local/builds/${safeJobName}`);
+            const time = process.hrtime();
+            const cacheName = await c.getUniqueCacheName(this.cwd, this.expandedVariables);
+            const cacheFolder = `${this.cwd}/.gitlab-ci-local/cache/${cacheName}`;
+            if (!await fs.pathExists(cacheFolder)) {
+                continue;
             }
-            return Utils.spawn(`docker cp ${folder}/. ${this._containerId}:/builds/${safeJobName}`);
-        };
+
+            await Mutex.waitForLock(cacheName);
```

Comment:
Should this not be a call to `await Mutex.exclusive(cacheName, async () => await this.copyIn(cacheFolder));` instead?
This just waits until no one has the mutex anymore and then carries on. What happens if someone grabs it right in the middle of the call to `copyIn` ?

Reply:
@FnuGk Hmmmm.... You are right...

Reply:
@FnuGk There you go